### PR TITLE
3-add-cart-system

### PR DESCRIPTION
**Pull request by wendy0402**

### CODE REVIEW

File: app/controllers/orders_controller.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  def no_cart
+    flash[:alert] = 'you have not put anything in the cart'
+    redirect_to root_path
+  end
+
+  def create
+    if current_user.present?
+      order = find_order
+      if order.blank?
+        params[:alert] = "This Order is not belongs to you"
```
Comment (adamnoto) [inline app/controllers/orders_controller.rb]: Wen, did you mean `flash[:alert]`? If so, why not just pass it as an argument to the `redirect_to` bellow

File: app/controllers/orders_controller.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    @orders = Order.for_user(current_user)
+  end
+
+  def show
+    @order = Order.find(params[:id])
+  end
+
+  def no_cart
+    flash[:alert] = 'you have not put anything in the cart'
+    redirect_to root_path
```
Comment (adamnoto) [inline app/controllers/orders_controller.rb:12]: Would you like to explain me, why this line of code doesn't include `alert` instead?

File: app/controllers/orders_controller.rb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+  def create
+    if current_user.present?
+      order = find_order
+      if order.blank?
+        params[:alert] = "This Order is not belongs to you"
+        redirect_to :back
+        return
+      end
+
+      if !params[:order_detail][:quantity].to_s.match(/\A[+-]?\d+?(\.\d+)?\Z/) || params[:order_detail][:quantity].to_i < 0
```
Comment (adamnoto) [inline app/controllers/orders_controller.rb]: this regex seems complex to me, why is that superior to just `/^\d+$/`

And `to_i` will always convert to an integer, so may as well remove the check for match.

File: app/controllers/orders_controller.rb (hunk 1 above)
Comment (wendy0402) [inline app/controllers/orders_controller.rb]: good point

File: app/controllers/orders_controller.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        flash[:alert] = 'quantity must be number and greater than 0'
+        redirect_to :back
+        return
```
Comment (adamnoto) [inline app/controllers/orders_controller.rb:28]: Seems like an unnecessary `return` to me

File: app/controllers/orders_controller.rb (hunk 2 above)
Comment (wendy0402) [inline app/controllers/orders_controller.rb:12]: it's already included dude, when you define flash[:alert]. still included

File: app/controllers/orders_controller.rb (hunk 3 above)
Comment (wendy0402) [inline app/controllers/orders_controller.rb]: do you have any better way to validate params to be number and must be greater than 0?

File: app/controllers/orders_controller.rb (hunk 4 above)
Comment (wendy0402) [inline app/controllers/orders_controller.rb:28]: when you do `redirect_to` and `render`, it's still run the other code, if it found another `redirect_to`, it will raise error. Furthermore since it's still running other code, `Order` record will be created, and it's detail too

File: app/controllers/orders_controller.rb [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+      end
+
+      book = Book.find(params[:order_detail][:book_id])
```
Comment (adamnoto) [inline app/controllers/orders_controller.rb:31]: In my opinion, new coder will be confused at `find` here, let's use `find_by` or `find_by_id` if that is doable.

Comment (wendy0402) [inline app/controllers/orders_controller.rb:31]: the idea is to show to the user, reason why someone use `find`, instead of `find_by` and `find_by_id`. Showing show page is a good use case for me

File: app/controllers/orders_controller.rb [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+      order_detail = OrderDetail.find_by(book: book, order: order)
+      success = false
+      if order_detail.present?
+        order_detail.quantity += params[:order_detail][:quantity].to_i
+        success = order_detail.save
+      else
+        order_detail = OrderDetail.create(book: book, quantity: params[:order_detail][:quantity], order: order)
```
Comment (adamnoto) [inline app/controllers/orders_controller.rb]: if code is over 80 characters, it'd be hard to put it in a book, as demonstrated by wrapping here. Maybe this:

``` ruby
OrderDetail.create(
  book: book,
  quantity: params[:order_detail][:quantity],
  order: order,
)
```

Comment (wendy0402) [inline app/controllers/orders_controller.rb]: sure

File: app/helpers/application_helper.rb [hunk 7]
```diff
@@ -1,2 +1,10 @@
 module ApplicationHelper
+  def money_amount(amount)
+    number_to_currency(amount, unit: 'Rp.')
+  end
+
+  def short_summary(words, length, link)
+    summary = words[0 .. length]
```
Comment (adamnoto) [inline app/helpers/application_helper.rb]: I think it'd be better using truncate:

``` ruby
"This book is amazing and cool".truncate(17, separator: ' ')
```

That will truncate per-word rather than per-character, and also will add "..."

File: app/models/user.rb [hunk 8]
```diff
@@ -1,4 +1,6 @@
 class User < ActiveRecord::Base
+
```
Comment (adamnoto) [inline app/models/user.rb]: seems like an unnecessary whitespace

File: app/views/home/index.html.erb [hunk 9]
```diff
@@ -1,26 +1,23 @@
 <div class='card-deck'>
-   <div class='card'>
-      <div class='card-block'>
-         <h4 class='card-title'>iPhone 5S</h4>
-         <h6 class='card-subtitle text-muted'>Rp. 8.999.999</h6>
+  <% @books.each do |book|%>
```
Comment (adamnoto) [inline app/views/home/index.html.erb]: for readability, please insert a single space after the last bar "|" before the "%"

File: app/views/home/index.html.erb [hunk 10]
```diff
@@ -1,26 +1,23 @@
...
+    <div class='card'>
+      <div class='card-block card-header background-white'>
+        <h5 class='card-title'><%= book.title%></h4>
```
Comment (adamnoto) [inline app/views/home/index.html.erb]: this also, need a whitespace after `.title` before before `%`

File: app/views/home/index.html.erb [hunk 11]
```diff
@@ -1,26 +1,23 @@
...
-            <div class='dropdown-menu'>
-               <a class='dropdown-item' href='#'>Action</a>
-               <a class='dropdown-item' href='#'>Another action</a>
-            </div>
-         </div>
-         <a href='#' class='btn btn-secondary card-link'>Quick checkout</a>
+        <p class='card-text card-summary'>
+          <%= short_summary(book.summary,100, '#') %>
+        </p>
+        <%= form_for @order_detail, url: orders_path do |f|%>
```
Comment (adamnoto) [inline app/views/home/index.html.erb]: this also, prior to `%`

File: app/views/orders/index.html.erb [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+  <table class="table table-hover table-responsive">
+    <thead>
+      <tr>
+        <td> order id</td>
+        <td> order date</td>
+        <td> status </td>
+        <td> action </td>
+      </tr>
+    </thead>
+    <% @orders.each do |order|%>
```
Comment (adamnoto) [inline app/views/orders/index.html.erb]: this also

File: app/views/orders/index.html.erb [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+      <tr>
+        <td> <%= order.id %></td>
+        <td> <%= order.created_at.strftime("%d %B %Y")%>
```
Comment (adamnoto) [inline app/views/orders/index.html.erb]: this too

File: app/views/orders/index.html.erb [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+        <td> <%= order.status %></td>
+        <td> <%= link_to 'details', order_path(order)%></td>
```
Comment (adamnoto) [inline app/views/orders/index.html.erb]: and this ^^

File: app/views/orders/show.html.erb [hunk 15]
```diff
@@ -0,0 +1,44 @@
+<div class="col-sm-12 col-sm-offset-2">
+  <table class="table table-hover table-stripped table-responsive">
+    <thead>
+      <tr>
+        <td colspan='2'>Buku</td>
+        <td>Harga Satuan</td>
+        <td>Kuantitas</td>
+      </tr>
+    </thead>
+    <% @order.order_details.each do |order_detail|%>
```
Comment (adamnoto) [inline app/views/orders/show.html.erb]: and this `%` also deserves a space after that bar, it feels crowded

File: app/views/orders/show.html.erb [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+          <%= order_detail.book.title %>
+        </td>
+
+        <td>
+          <%= money_amount(order_detail.unit_price) %>
+        </td>
+
+        <td>
+          <div class="inline-block">
+            <%= form_for order_detail, url: order_order_detail_path(order_id: @order.id, id: order_detail.id) do |f| %>
```
Comment (adamnoto) [inline app/views/orders/show.html.erb]: maybe in partial

File: app/controllers/orders_controller.rb (hunk 2 above)
Comment (adamnoto) [inline app/controllers/orders_controller.rb:12]: OK so you prefer `flash[:alert]`, that's fine too. I feel like it is shorter if combined

File: app/controllers/orders_controller.rb (hunk 3 above)
Comment (adamnoto) [inline app/controllers/orders_controller.rb]:
``` ruby
quantity = params[:order_detail][:quantity]
if quantity =~ /^\d+$/ && quantity.to_i > 0
   # bla bla bla
end
```

Or I think it is better:

``` ruby
quantity = params[:order_detail][:quantity].to_i
if quantity > 0
  # bla bla bla
end
```

Because whatever it is, `to_i` will always convert a String to an Integer right? So checking if it contains other than number is unnecessary, because by then it will be converted to 0.

File: app/controllers/orders_controller.rb (hunk 4 above)
Comment (adamnoto) [inline app/controllers/orders_controller.rb:28]: I see. This is `create`, sorry.

File: app/controllers/orders_controller.rb (hunk 5 above)
Comment (adamnoto) [inline app/controllers/orders_controller.rb:31]: And what is that reason?

File: app/views/orders/show.html.erb (hunk 16 above)
Comment (wendy0402) [inline app/views/orders/show.html.erb]: the form is not complex

Comment (adamnoto) [inline app/views/orders/show.html.erb]: It is not, but the book wouldn't be able to contain it (> 80 characters) properly, because of the word wrap.

In my personal capacity as reader, I don't like it when opening a programming book and the line is wrapped to the next line, it makes it hard to read.

File: app/controllers/orders_controller.rb (hunk 5 above)
Comment (wendy0402) [inline app/controllers/orders_controller.rb:31]: when you try to open a show page, and that order doesn't exists its expected to not give you anything.

File: app/controllers/orders_controller.rb (hunk 3 above)
Comment (wendy0402) [inline app/controllers/orders_controller.rb]: I'll use the first one,
second one: try `12e1`.to_i, and you will know why

File: app/views/orders/show.html.erb (hunk 16 above)
Comment (wendy0402) [inline app/views/orders/show.html.erb]: ok

File: app/controllers/orders_controller.rb (hunk 5 above)
Comment (adamnoto) [inline app/controllers/orders_controller.rb:31]: no I mean, why use `find`? instead of `find_by` or `find_by_something`

File: app/controllers/orders_controller.rb (hunk 3 above)
Comment (adamnoto) [inline app/controllers/orders_controller.rb]: Hahaha OK YOU ARE SO AWESOME

Ok then, use `=~ /^d+$/` is enough IMO

Comment (adamnoto) [inline app/controllers/orders_controller.rb]: whoisawesome

File: app/controllers/orders_controller.rb (hunk 5 above)
Comment (wendy0402) [inline app/controllers/orders_controller.rb:31]: that's the reason

File: app/controllers/orders_controller.rb [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+    if current_user.present?
+      order = find_order
+      if order.blank?
+        flash[:alert] = "This Order is not belongs to you"
+        redirect_to :back
+        return
+      end
+
+      quantity = params[:order_detail][:quantity]
+      if quantity !=~ /^\d+$/ && quantity.to_i < 0
```
Comment (adamnoto) [inline app/controllers/orders_controller.rb]: Don't use '!' it will make it evaluates to true no matter what. Either

``` ruby
!!(quantity =~ /^\d+$/)
```

Which is cryptic, or:

``` ruby
(quantity =~ /^\d+$/).nil?
```

File: app/controllers/orders_controller.rb (hunk 3 above)
Comment (wendy0402) [inline app/controllers/orders_controller.rb]: no you can't, 0 is not allowed. that regex allow 0
